### ISSUE TITLE
 Chat - Height of edit field is different for mWeb and native app @thesahindia

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#11694]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Launch the app
2. Navigate to any chat
3. Send a message
4. Long press on it > Edit
5. Keep pressing enter and check the max height

## Expected Result:
The height should be consistent

## Actual Result:
The height isn't consistent

## Workaround:
Unknown

## Platform:

Where is this issue occurring?

- Web
- iOS
- Android
- Desktop App
- Mobile Web

**Version Number:** 1.2.12.2

**Reproducible in staging?:** Yes

**Reproducible in production?:** Yes

**Email or phone of affected tester (no customers):** any

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation
![Screenshot_20220926_182516](https://user-images.githubusercontent.com/93399543/194785714-b5f4ee7c-73f4-4063-bf68-8517cbf811c9.jpg)
![Screenshot_20220926_182452](https://user-images.githubusercontent.com/93399543/194785715-ed791a76-d7e5-4509-b429-550abb048e6a.jpg)

**Expensify/Expensify Issue URL:** 

**Issue reported by:** @thesahindia

**Slack conversation:** https://expensify.slack.com/archives/C01GTK53T8Q/p1664197609260269

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by Puneet-here** [comment]
### Proposal

We need to keep the max lines 6 for isSmallScreenWidth and 16 when it's false

https://github.com/Expensify/App/blob/f84651934726d25e720047839c4ed750eea111b6/src/pages/home/report/ReportActionItemMessageEdit.js#L201

``` maxLines={this.props.isSmallScreenWidth ? 6 : 16} ```
we can keep add them in CONST.js

**Comment by strepanier03** [comment]
Triage notes:

**This issue is repeatable.**

Tested this on my Android using the new.expensify mobile app. 

I was able to have a height much greater than 6 or 16, it seemed near infinite and I could keep increasing the height repeatedly.

<img width="1045" alt="2022-10-10_13-22-47" src="https://user-images.githubusercontent.com/10925636/194946493-34334a5c-e436-4199-8114-b92466221bba.png">

**Comment by strepanier03** [comment]
I think even with a proposal, this should go to Eng according ot the linked SO. There's multiple moving parts to our transition to BZ and I think this Triage process will go away in it's current form soon, a DD was just released for reviews regarding it. 

Based on this [thread](https://expensify.slack.com/archives/C01SKUP7QR0/p1665429569726079?thread_ts=1665428390.597069&cid=C01SKUP7QR0) in Slack, I think I should still triage as normal for now, with this going to eng who can put the external label on it as needed??

**Comment by roryabraham** [comment]
@Puneet-here's [proposal](https://github.com/Expensify/App/issues/11694#issuecomment-1272868320) looks good to me, but I'll let @rushatgabhane buddy-check it. I'm not sure if we need to check `canUseTouchScreen` or if just screen width is sufficient here.

**Comment by Puneet-here** [comment]
> @Puneet-here question: In message edit, why is the max line height 6 for native? Shouldn't it be 16 because we're setting it over here?

It's probably because of the styling

> Also, I'm curious why we should use `canUseTouchScreen()`, is it because it's mWeb only?

I think it's good to use `canUseTouchScreen()` and `isSmallScreenWidth ` because this way the height of composer won't change for desktop when we decrease the screen size

**Comment by sonialiap** [comment]
[Job is posted to Upwork](https://www.upwork.com/ab/applicants/1583210411543572480/job-details)

**Comment by Puneet-here** [comment]
> @Puneet-here I think we need to dive deeper then. Like which specific style is changing maxLine behavior for native only?

I was a little bit busy, I will check this today.

**Comment by Puneet-here** [comment]
Apologies, I just realised that I had closed the tab without really pressing the comment button earlier.

So we have platform specific file for composer and because we are passing maxHeight, passing maxLines has no effect on it
https://github.com/Expensify/App/blob/83676c1036299ffecbf82f4b829a7f8cbd741cd7/src/components/Composer/index.ios.js#L104

**Comment by roryabraham** [comment]
> @roryabraham do you think a cleaner solution would be to implement maxLines for native using maxHeight?

I think a cleaner solution would probably be to implement the `maxLines` prop in React Native for iOS and use `maxLines` consistently across platforms.

**Comment by roryabraham** [comment]
Overall I feel like:

1. The platform differences of the `Composer` have diverged more than they should have, such that the differences between the platforms aren't clear
1. There is a lot of custom logic in the `Composer` component, much of which is probably intentional, but a lot of the context behind that logic is not very transparent. Some automated tests which cover the behaviors that we expect would go a long way into preserving the intention of the component.
1. What is even happening with `isComposerFullSize`?
    - Why do we have [three separate values](https://github.com/Expensify/App/blob/18ff316bf003fddfd68868f5a771b5f73f8f99ce/src/pages/home/report/ReportActionCompose.js#L674-L676) for `isFullComposerAvailable`, `setIsFullComposerAvailable`, and `isComposerFullSize`?
    - `isComposerFullSize` is actually persisted in Onyx on a report level: https://github.com/Expensify/App/blob/0e147426759c9cf4878479f3ffeb69e54249aeb6/src/pages/home/ReportScreen.js#L341-L343, is that something we actually want to persist between sessions or are we just using that as a way to manage UI state?
    - Do we actually need a separate value for `isFullComposerAvailable`? It seems like it's just shadowing `isComposerFullSize`.
1. It seems like all of this boils down to `maxLines`, which we could use directly across all platforms if it was implemented in React Native iOS.
1. Why is [maxLines smaller for small screens](https://github.com/Expensify/App/blob/18ff316bf003fddfd68868f5a771b5f73f8f99ce/src/pages/home/report/ReportActionCompose.js#L306)? Is this because we have a software keyboard that takes up screen height? If so, why don't we determine `maxLines` based on whether the software keyboard is open, rather than based on platform or screen width?

I think the wisest course of action would be:

- To do a Full Throttle (FT) or Slack meeting for the Composer component across platforms, to identify the behaviors we expect.
- (optional, but ideal) Write some automated tests to cover the expected behavior
- Update the Composer component to address this issue and some of the above concerns with as simple and cross-platform an implementation as possible, implementing `maxLines` in React Native for iOS if needed.

What do people think?

**Comment by roryabraham** [comment]
Didn't have time to do a deep-dive on this myself yet, but I did take the opportunity to do one pass on cleanup of the Composer component in [this PR](https://github.com/Expensify/App/pull/12918). It's not much, but it's an easy win and a step in the right direction.

Long-term for this specific bug I think we probably want to implement `maxLines` on iOS as well as Android in RN.

**Comment by sonialiap** [comment]
Price doubled 🙌

**Comment by roryabraham** [comment]
[This PR](https://github.com/Expensify/App/pull/12918) turned out to be not quite ready yet. An [upstream fix](https://github.com/Expensify/react-native-web/pull/9) is needed, but then it will be ready

**Comment by tjferriss** [comment]
Following instructions for the [weekly update chore](https://stackoverflow.com/c/expensify/questions/8208):

@roryabraham this is one of the oldest issues in the /App repo. To help us clear out the large backlog of bugs, can you:

- Decide whether any proposals currently meet our guidelines and can be approved as-is
- For any that can't, please take this issue internal and treat it as one of your highest priorities
- If you have any questions, don't hesitate to start a discussion in #bug-zero

**Comment by trjExpensify** [comment]
Ah, just commented [here in thread](https://expensify.slack.com/archives/C01GTK53T8Q/p1669227930360779?thread_ts=1669139010.581789&cid=C01GTK53T8Q). I’m changing the label to `Internal` and removing `Help wanted`.

**Comment by roryabraham** [comment]
> Ah, just commented [here in thread](https://expensify.slack.com/archives/C01GTK53T8Q/p1669227930360779?thread_ts=1669139010.581789&cid=C01GTK53T8Q). I’m changing the label to Internal and removing Help wanted.

Sorry, I think there was a miscommunication here @trjExpensify. The PR I put up was a step in the right direction, but not a solution to this issue.

Also the [upstream fix](https://github.com/Expensify/react-native-web/pull/9) was merged in our fork, but in the actual react-native-web upstream, the maintainer has stopped accepting pull requests in the VirtualizedList components. I think the plan is to move VirtualizedList into a separate package of React Native. We need to have a conversation about what to do here.

**Comment by JmillsExpensify** [comment]
Going to float this one to Margelo today.

**Comment by roryabraham** [comment]
Still waiting for proposals that fix this by implementing `numberOfLines` for React Native TextInput on iOS ([currently an Android-only prop](https://reactnative.dev/docs/textinput#numberoflines-android)), such that it can be used consistently across platforms.

**Comment by sonialiap** [comment]
Price doubled 🥳

**Comment by sonialiap** [comment]
waiting for proposals

**Comment by roryabraham** [comment]
We're getting some help from @Szymon20000 on the upstream `numberOfLines` prop.

**Comment by Szymon20000** [comment]
I added native support for `numberOfLines` and `maximumNumberOfLines` on iOS. Here is a pr where you can test it. 
https://github.com/Expensify/App/pull/13367

**Comment by JmillsExpensify** [comment]
@sonialiap Please note that @Szymon20000 qualifies for the bug bounty. Thanks!

**Comment by Szymon20000** [comment]
https://github.com/Expensify/react-native/pull/37 
Here is the pr introducing above 2 props to Expensify/react-native fork.

**Comment by roryabraham** [comment]
> @sonialiap Please note that @Szymon20000 qualifies for the bug bounty. Thanks!

Sorry for the multiple miscommunications here @JmillsExpensify @sonialiap et al, but I think @Szymon20000 should just bill us separately for his work in the React Native fork. Completing https://github.com/Expensify/react-native/pull/37 is a _prerequisite_ of fixing this bug but will not fix it outright, so we'll want to hire someone separately to _use_ the new feature to fix this bug. Hopefully that makes sense. Either that or just hire @Szymon20000 to take it over the finish line with an E/App PR, which I'd also be fine with (and then he'd qualify for the bounty on top of his hourly rate).

**Comment by trjExpensify** [comment]
👋 Should https://github.com/Expensify/react-native/pull/37 be in review, not a draft?

**Comment by Szymon20000** [comment]
After long hours of uneven struggle, I found a strange bug in NSContainer, which turned out to be visible not only in Fabric.
However, in paper architecture it was only visible in text. Long story short when you set some properties in a wrong order than adding new lines causes text/text input to grow even though we set maximumNumberOfLines. Here is the commit: https://github.com/Expensify/react-native/pull/37/commits/5db9c6f77c265c7a5bed0e4c22b3a7fb5706f6e1

I will clean up the pr on Monday morning and would be very happy if someone could test it.

**Comment by trjExpensify** [comment]
Nice! Thanks for getting that up, Szymon! @roryabraham @rushatgabhane are you guys taking the reviews and testing on that? Let's get you added to the PR as reviewers now if so! 

Also, where did we land on [this comment](https://github.com/Expensify/App/issues/11694#issuecomment-1344968442)? @Szymon20000 did you want to take the App PR as well? Given the age of this issue, we should find an `internal` volunteer for it if not @roryabraham. Either way, this issue is held on the Expensify/react-native PR which is a prerequisite to fixing the bug this issue was created for. I'm updating the title to reflect the dependency.

**Comment by roryabraham** [comment]
Left a comment on the react-native PR

**Comment by Szymon20000** [comment]
> Nice! Thanks for getting that up, Szymon! @roryabraham @rushatgabhane are you guys taking the reviews and testing on that? Let's get you added to the PR as reviewers now if so!

> Also, where did we land on https://github.com/Expensify/App/issues/11694#issuecomment-1344968442? @Szymon20000 did you want to take the App PR as well? Given the age of this issue, we should find an internal volunteer for it if not @roryabraham. Either way, this issue is held on the Expensify/react-native PR which is a prerequisite to fixing the bug this issue was created for. I'm updating the title to reflect the dependency.

Sure, I can open a pr fixing this issue in the App repo once the RN pr is merged. It's 5min fix then.

**Comment by Szymon20000** [comment]
Just switched the base branch of RN pr to [Expensify-0.71.0-alpha1](https://github.com/Expensify/react-native/tree/Expensify-0.71.0-alpha1) branch + opened a twin pull-request in react-native repo.

**Comment by sonialiap** [comment]
Holidays, people out

**Comment by sonialiap** [comment]
Holidays, people have been out

**Comment by sonialiap** [comment]
I believe people are still OOO

**Comment by JmillsExpensify** [comment]
Latest update in Slack. On that note though, where's the PR for this issue? We should associate the two together.

**Comment by JmillsExpensify** [comment]
Oh sorry this one is in a private room, apologies. Not sure if there's a good reason for that, so thanks for the reminder.

**Comment by roryabraham** [comment]
Sorry folks, this is currently blocked on https://github.com/Expensify/react-native/pull/37, which is shadowing the upstream https://github.com/facebook/react-native/pull/35703. Between the holidays and some higher-priority projects, I have not found the time to test and review https://github.com/Expensify/react-native/pull/37, so that's the next step here.

**Comment by roryabraham** [comment]
Same as last week unfortunately ☹️

**Comment by sonialiap** [comment]
I believe we're still waiting on https://github.com/Expensify/react-native/pull/37

**Comment by sonialiap** [comment]
Still waiting on https://github.com/Expensify/react-native/pull/37

Going to move this to Weekly

**Comment by sonialiap** [comment]
I'm OOO Feb 2-11. I'm guessing this won't move forward before I'm back, but just FYI

**Comment by sonialiap** [comment]
On hold

**Comment by sonialiap** [comment]
https://github.com/Expensify/react-native/pull/37 still open, this one's still on hold

**Comment by sonialiap** [comment]
https://github.com/Expensify/react-native/pull/37 still open, this one's still on hold

**Comment by sonialiap** [comment]
Same ol story, waiting for https://github.com/Expensify/react-native/pull/37

**Comment by roryabraham** [comment]
https://github.com/Expensify/react-native/pull/37 was merged and published in [v0.71.2-alpha.3 of our React Native fork](https://github.com/Expensify/react-native/releases/tag/v0.71.2-alpha.3). Now this will need to be on HOLD for https://github.com/Expensify/App/pull/15130

**Comment by sonialiap** [comment]
Still on hold

**Comment by sonialiap** [comment]
@roryabraham https://github.com/Expensify/App/pull/15130 was merged, is this off-Hold now?

**Comment by chrispader** [comment]
**Proposal**

Change the `maxLines` prop on the `Composer` in "ReportActionItemMessageEdit" as @Puneet-here already proposed [here](https://github.com/Expensify/App/issues/11694#issuecomment-1272868320).

Additionally use the `maximumNumberOfLines` prop in the composer on `RNTextInput` on iOS and Android instead of `maxHeight`. 

I've just started working on a [PR](https://github.com/Expensify/App/pull/18214/files#diff-4b0b786d38a265539a2fa5a0928938a16c50018686546cc17a412261682385eb) for fixing this issue, by using the `numberOfLines` and `maxNumberOfLines` props, @Szymon20000 recently introduced to the RN-fork [here](https://github.com/Expensify/react-native/pull/37).

**Comment by Naveen3Singh** [comment]
The merged PR, located at https://github.com/Expensify/App/pull/18214, has introduced two new issues that need attention:

**Issue: Text field expansion** - The expansion of the text field to its full-size results in only one visible line of text as shown in the screenshot instead of the expected behavior (it should display multiple lines of text, but it currently shows only one line)
  <img src="https://github.com/Expensify/App/assets/36371707/1948de5e-13a8-49a9-a250-72bc3336c7e8" width=264 />

**Issue: App crash when collapsing** - The app crashes frequently when attempting to close the expanded view using the collapse button. Check out the video below :point_down: 


https://github.com/Expensify/App/assets/36371707/d38c0ef2-2478-43a0-8627-1c7bc3032936



@sonialiap, @rushatgabhane, @roryabraham I'm new here but as per the contribution guidelines, I believe `DeployBlockerCash` label needs to be added. 

**Off-topic:** I tried sending a mail to contributors@expensify.com with `Slack Channel Invites` as the subject but haven't got any invitation yet, can anyone one of you help me with that?

**Comment by Naveen3Singh** [comment]
Contributor details
Your Expensify account email: snaveen935@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~014e26b8316fe30a5f

**Comment by Naveen3Singh** [comment]
[Updated comment with steps to reproduce both issues.](https://github.com/Expensify/App/issues/11694?notification_referrer_id=NT_kwDOAir8-7M0NTcwOTc2MDgxOjM2MzcxNzA3#issuecomment-1560520670)

**Comment by situchan** [comment]
The [PR](https://github.com/Expensify/App/pull/18214) which fixes this issue caused regression.


I have a solution but I will let assigned folks handle as it's within regression period.

**Comment by chrispader** [comment]
Fixed [this regression](https://github.com/Expensify/App/issues/11694#issuecomment-1560520670) [in this PR](https://github.com/Expensify/App/pull/19809)

Seems there is a bug with the recently introduced `maxNumberOfLines` prop in the RN-fork. @Szymon20000 is working on fixing it!

cc @rushatgabhane

**Comment by Szymon20000** [comment]
Hey guys!
I just reproduced it in rn-tester repo on my branch. Thanks for finding it. 
Will work on it tomorrow. Let's for now use "inf" instead of undefined for full-screen.

**Comment by Szymon20000** [comment]
I'm no longer sure if it's a bug. The same thing would happen if we would use numberOfLines before my changes. 
Basically ReactNative implements single and multiline components with the same component. In some cases when we use
maxNumberOfLines to undefined after it was set before to something else we want to set it to 1 in some to inf.

**Comment by Szymon20000** [comment]
<img width="855" alt="Screenshot 2023-05-31 at 09 45 35" src="https://github.com/Expensify/App/assets/12784455/1ab39b6d-43a1-49a5-aa5f-2f7d80ba8cfc">
So basically when we set the property to something different than undefined and then to undefined the default value will be used. It's not clear for me what should be the default value as it depends on other props like `multiline`.

**Comment by sonialiap** [comment]
Not ready for payment because regression was found and needs fixing 🛠️

**Comment by sonialiap** [comment]
Whoops, this didn't automatically change to Daily. I'm clocking off now but I'll get the payments set up tomorrow morning!

**Comment by sonialiap** [comment]
Regression test request submitted. 

@rushatgabhane feel free to close once this has been payed out

**Comment by roryabraham** [comment]
@rushatgabhane does not have triage or write access, so @rushatgabhane @JmillsExpensify @anmurali once payment is done here please close this issue. Going to make this weekly in the meantime.

**Comment by sonialiap** [comment]
Thanks for the confirmation, closing out

**Comment by naveensingh** [comment]
> I will request the money on new dot.

what do you mean by that? I'm new here.

**Comment by naveensingh** [comment]
@rushatgabhane what do I need to do here? 
I already provided my contributor details when I reported the regression above but I changed my username from @naveen3singh to @naveensingh. Does @melvin uses Github username as the contributer id?

**Comment by naveensingh** [comment]
Contributor details
Your Expensify account email: snaveen935@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~014e26b8316fe30a5f

**Comment by naveensingh** [comment]
second time's the charm :smile:

**Comment by naveensingh** [comment]
@sonialiap, @rushatgabhane, @roryabraham am I supposed to be compensated for https://github.com/Expensify/App/issues/11694#issuecomment-1560520670? If yes, do I need to do anything from my end? Contribution guidelines aren't very clear on reporting regressions.

**Comment by JmillsExpensify** [comment]
Awaiting payment summary for issue reporting.

**Comment by JmillsExpensify** [comment]
Reached out to @sonialiap via NewDot.

**Comment by JmillsExpensify** [comment]
$250 payment approved for @thesahindia based on summary above.
